### PR TITLE
Comment: use `@combined` to create just one tree per buffer

### PR DESCRIPTION
**Pull request by stsewd**
There is no need to create a tree per line/block for comments.
Should fix https://github.com/nvim-treesitter/nvim-treesitter/issues/1251

### CODE REVIEW

File: queries/ecma/injections.scm [hunk 1]
```diff
@@ -1,5 +1,5 @@
 (comment) @jsdoc
```
Comment (theHamsta) [inline queries/ecma/injections.scm:1]: Doesn't this also apply to us jsdoc? Are really both parsers required?

Comment (stsewd) [reply queries/ecma/injections.scm:1]: Think it was decided to use both, jsdoc and comment for the injection. Don't know if jsdoc should be combined.

Comment (theHamsta) [reply queries/ecma/injections.scm:1]: jsdocs has the same performance problem on my machine

Comment (stsewd) [reply queries/ecma/injections.scm]: Changed

**Comment by theHamsta** [comment]
@steelsojka what do you think about combined jsdoc as quickfix?

**Review by theHamsta — APPROVED** [review]

**Comment by vigoux** [comment]
I am quite strongly against this change, as it is just a workaround.
What really need to be done is improve the performances of the language tree, and not hacks.

**Comment by theHamsta** [comment]
The comment parser is a hack. We could also remove it if it's not working properly for now which is the solution that I had chosen. When it's not ready it's not ready. Just assuming that people want to delete lines this week.

**Comment by steelsojka** [comment]
@theHamsta What is the actually issue? Is it the treesitter query taking to long or the parser itself?

**Comment by steelsojka** [comment]
I don't think this is a hack as it is working as intended. The queries are just incorrect and not using a combined context when it should be. More of the issue is most likely in the amount of trees generated and not the language tree itself. 100 comments produces 100 iterations on each highlight. No matter how you try and optimize it you still end up with the same number of iterations.

**Comment by stsewd** [comment]
I don't this is a hack, it does make sense to have only one tree per buffer instead of one per line/block.

This change doesn't change the fact that having multiple trees is slow and that's something that should be fixed too.

**Comment by theHamsta** [comment]
I don't know the exact issue but it appears that the Lua and query overhead gets very big when we have one parser per line in case of line comments (a library file can easily have . I don't know how combined can help with that but I imagine that @stsewd and the user did some experiments.

**Comment by steelsojka** [comment]
I think having multiple trees (a lot or trees) is going to be slow no matter how you implement it. We could probably speed up the updating of those trees on edits, but you still have the same amount of trees and iterations during highlighting, etc... Even after we make some improvements we would still want this to be `combined`, so I don't consider it a hack.

**Comment by stsewd** [comment]
The problem I saw is when you create several trees by commenting several lines at the same time, I haven't see any slowness when creating a single comment in a file with no comments. It's also slow when you uncomment several lines don't know if that's because de plugin uncomments line by line or destroying all those trees is slow, I can do a quick test in some minutes.

**Comment by steelsojka** [comment]
@theHamsta I think this is just a suspected problem.  I don't think we've pin pointed exactly where it is slow.

**Comment by steelsojka** [comment]
@stsewd I've noticed this with commenting plugins that do it line by line. I believe treesitter parses on each line that gets commented. Our reparse is on the bytes callback, so maybe this is a bit to eager. I wonder if there's a callback we can have that would call at the end of the loop after any text edits would be completed but prior to the rendering and highlighting.

**Comment by stsewd** [comment]
Just checked, commenting/uncomenting without using the plugin is fast, looks like when hitting undo (u) is also done line by line, that's something that neovim could improve I think.

**Comment by steelsojka** [comment]
I've investigated this and I think this is the correct solution. It doesn't matter that it's done line by line. It's more to the fact that each line is an injected region and each region has to be parsed. So, 1000 comments means 1000 parse calls. Currently there's no incremental parsing for injected languages, but it won't matter if you are just commenting out lines since it hasn't been parsed yet. I'll look at incremental parsing for injected languages, but it won't matter on initial parse.

**Comment by steelsojka** [comment]
@theHamsta We should do this for jsdoc as well. Jsdoc is alright funky because it will error if you comment out a line like `// function () {}` We should make that injection a little bit smarter.

**Review by steelsojka — APPROVED** [review]

**Pull request merged**